[AutoDiff] Enable ownership for JVP/VJP SILGen thunks.

dan-zheng opened the pull request:
Enable ownership for JVP/VJP SILGen thunks.
Ownership verification ensures that these thunks do not leak memory.

Delete redundant helper `forwardFunctionArgumentsConvertingOwnership`,
which is no longer necessary.

Various thunk leak-checking tests already exist in
`test/AutoDiff/leakchecking.swift`.

Resolves [TF-703](https://bugs.swift.org/browse/TF-703): all AutoDiff SILGen thunks now use ownership.

dan-zheng committed:
[AutoDiff] Enable ownership for JVP/VJP SILGen thunks.

Enable ownership for JVP/VJP SILGen thunks.
Ownership verification ensures that these thunks do not leak memory.

Delete redundant helper `forwardFunctionArgumentsConvertingOwnership`,
which is no longer necessary.

Various thunk leak-checking tests already exist in
test/AutoDiff/leakchecking.swift.

Resolves TF-703: all AutoDiff SILGen thunks now use ownership.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Minor edits.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -3760,52 +3709,43 @@ SILGenModule::getOrCreateAutoDiffAssociatedFunctionThunk(
   // Otherwise, apply reabstraction/self reordering thunk to linear map.
+  SmallVector<SILValue, 8> directResults;
+  extractAllElements(apply, loc, thunkSGF.B, directResults);
+  auto linearMap = ManagedValue::forBorrowedObjectRValue(directResults.back());
```
This cannot be a borrowed rvalue. It is owned because it's a function call's result.

dan-zheng replied:
That makes sense, but I'm not sure what API to use instead. What do you recommend?

`linearMap` is known to be always consumed by `partial_apply`, so no cleanup is necessary.

We could use `ManagedValue::forOwnedObjectRValue` (which takes a `CleanupHandle`) and then disable the cleanup, but I'm not sure how to construct a `CleanupHandle`.

dan-zheng replied:
Changed to `SILGenFunction::emitManagedRValueWithCleanup` instead in https://github.com/apple/swift/pull/27393/commits/6b770a32de1e385359175322981e6b381e8dd3ef. Functionally identically here but more semantically accurate.

rxwei replied:
`emitManagedRValueWithCleanup` is the right API to use.

dan-zheng committed:
Use semantically correct `ManagedValue` construct.

`ManagedValue::forBorrowedObjectRValue` is inaccurate because `linearMap`
is actually an owned value.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
